File names in qf window can't get reduced (GVim), absolute paths and relative paths mix together and mess up qflist.

GVim: 8.0.1733
System: Windows 10

It seems like quickfix window will reduce file path name to be relative to current directory, not all the file names get reduced, though they are in the same directory.

![quickfix1](https://user-images.githubusercontent.com/3035071/39287337-d47322c6-4955-11e8-9ca2-14211a3cf131.png)

This is a simple `grep` output, all the files locate in the the current directory, some of them get reduced to become a relative path. while some of them keep absolute path name.

Relative path names and absolute path names from the same directory totally mess up the quickfix window. Any clue or option the fix this ?

If I add:

```VimL
exec (haslocaldir()? "lcd ." : "cd .")
```
before `caddexpr` in the job callback, the first line of the same file still uses absolute name, but the following lines of the same file use relative names:

```text
D:\acm\github\vim\site\calendar.vim|69| "     1.9  : bug fix, use nnoremap.
site\calendar.vim|384| nnoremap <silent> <Plug>CalendarV :cal Calendar(0)<CR>
site\calendar.vim|385| nnoremap <silent> <Plug>CalendarH :cal Calendar(1)<CR>
site\calendar.vim|1259| execute 'nnoremap <silent> <buffer> q :close<bar>wincmd p<cr>'
site\calendar.vim|1261| execute 'nnoremap <silent> <buffer> <Plug>CalendarDoAction  :call <SID>CalendarDoAction()<cr>'
site\calendar.vim|1262| execute 'nnoremap <silent> <buffer> <Plug>CalendarDoAction  :call <SID>CalendarDoAction()<cr>'
site\calendar.vim|1263| execute 'nnoremap <silent> <buffer> <Plug>CalendarGotoToday :call Calendar(b:CalendarDir)<cr>'
site\calendar.vim|1264| execute 'nnoremap <silent> <buffer> <Plug>CalendarShowHelp  :call <SID>CalendarHelp()<cr>'
site\calendar.vim|1265| execute 'nnoremap <silent> <buffer> <Plug>CalendarReDisplay :call Calendar(' . a:dir . ',' . a:vyear . ',' . a:vmnth . ')<cr>'
site\calendar.vim|1268| execute 'nnoremap <silent> <buffer> <Plug>CalendarGotoPrevMonth :call <SID>CalendarDoAction("<' . pnav . '")<cr>'
site\calendar.vim|1269| execute 'nnoremap <silent> <buffer> <Plug>CalendarGotoNextMonth :call <SID>CalendarDoAction("' . nnav . '>")<cr>'
site\calendar.vim|1270| execute 'nnoremap <silent> <buffer> <Plug>CalendarGotoPrevYear  :call Calendar('.a:dir.','.(a:vyear-1).','.a:vmnth.')<cr>'
site\calendar.vim|1271| execute 'nnoremap <silent> <buffer> <Plug>CalendarGotoNextYear  :call Calendar('.a:dir.','.(a:vyear+1).','.a:vmnth.')<cr>'
D:\acm\github\vim\site\dash.vim|26| noremap <script> <unique> <Plug>DashSearch <SID>DashSearch
site\dash.vim|27| noremap <SID>DashSearch :Dash<CR>
site\dash.vim|28| noremap <script> <unique> <Plug>DashGlobalSearch <SID>DashGlobalSearch
site\dash.vim|29| noremap <SID>DashGlobalSearch :Dash!<CR>
D:\acm\github\vim\site\echofunc.vim|473| inoremap <silent> <buffer>  (   (<C-R>=EchoFunc()<CR>
site\echofunc.vim|476| let map = maparg("(", "i", 0, 1)['noremap'] ? "inoremap" : "imap"
site\echofunc.vim|484| inoremap <silent> <buffer>  )   )<C-R>=EchoFuncClear()<CR>
site\echofunc.vim|487| let map = maparg(")", "i", 0, 1)['noremap'] ? "inoremap" : "imap"
site\echofunc.vim|495| exec 'inoremap <silent> <buffer> ' . g:EchoFuncKeyNext . ' <c-r>=EchoFuncN()<cr>'
site\echofunc.vim|496| exec 'inoremap <silent> <buffer> ' . g:EchoFuncKeyPrev . ' <c-r>=EchoFuncP()<cr>'
site\echofunc.vim|556| execute (b:maparg_left['noremap'] ? "inoremap" : "imap").' '
site\echofunc.vim|565| execute (b:maparg_right['noremap'] ? "inoremap" : "imap").' '
D:\acm\github\vim\site\indent-object.vim|28| onoremap <silent>ai :<C-u>cal <Sid>HandleTextObjectMapping(0, 0, 0, [line("."), line("."), col("."), col(".")])<CR>
site\indent-object.vim|29| onoremap <silent>ii :<C-u>cal <Sid>HandleTextObjectMapping(1, 0, 0, [line("."), line("."), col("."), col(".")])<CR>
site\indent-object.vim|30| vnoremap <silent>ai :<C-u>cal <Sid>HandleTextObjectMapping(0, 0, 1, [line("'<"), line("'>"), col("'<"), col("'>")])<CR><Esc>gv
site\indent-object.vim|31| vnoremap <silent>ii :<C-u>cal <Sid>HandleTextObjectMapping(1, 0, 1, [line("'<"), line("'>"), col("'<"), col("'>")])<CR><Esc>gv
site\indent-object.vim|34| onoremap <silent>aI :<C-u>cal <Sid>HandleTextObjectMapping(0, 1, 0, [line("."), line("."), col("."), col(".")])<CR>
site\indent-object.vim|35| onoremap <silent>iI :<C-u>cal <Sid>HandleTextObjectMapping(1, 1, 0, [line("."), line("."), col("."), col(".")])<CR>
site\indent-object.vim|36| vnoremap <silent>aI :<C-u>cal <Sid>HandleTextObjectMapping(0, 1, 1, [line("'<"), line("'>"), col("'<"), col("'>")])<CR><Esc>gv
site\indent-object.vim|37| vnoremap <silent>iI :<C-u>cal <Sid>HandleTextObjectMapping(1, 1, 1, [line("'<"), line("'>"), col("'<"), col("'>")])<CR><Esc>gv
D:\acm\github\vim\site\taglist.vim|1631| nnoremap <buffer> <silent> <CR>
site\taglist.vim|1633| nnoremap <buffer> <silent> o
site\taglist.vim|1635| nnoremap <buffer> <silent> p
site\taglist.vim|1637| nnoremap <buffer> <silent> P
site\taglist.vim|1640| nnoremap <buffer> <silent> t
```

that seems flaky.

IDEs display absolute path names in their output window, here is the output from visual studio:

```text
1>MainTest.cpp
1>e:\lesson\collection\contextswitch\maintest.cpp(87): error C2065: 'x': undeclared identifier
1>e:\lesson\collection\contextswitch\maintest.cpp(88): error C2065: 'y': undeclared identifier
1>e:\lesson\collection\contextswitch\maintest.cpp(89): error C2065: 'z': undeclared identifier
1>CtxWin64.cpp
```

Nobody uses a 80x25 character display now, is it possible to just display absolute names in quickfix window ? or to resolve every file path before adding to quickfix ?

--- a/src/nvim/fileio.c
+++ b/src/nvim/fileio.c
@@ -4312,7 +4312,7 @@ static int make_bom(char_u *buf, char_u *name)
 }
 
 /*
- * Shorten filenames for all buffers.
+ * Shorten filename of a buffer.
  * When "force" is TRUE: Use full path from now on for files currently being
  * edited, both for file name and swap file name.  Try to shorten the file
  * names a bit, if safe to do so.
@@ -4320,29 +4320,38 @@ static int make_bom(char_u *buf, char_u *name)
  * For buffers that have buftype "nofile" or "scratch": never change the file
  * name.
  */
+void shorten_buf_fname(buf_T *buf, char_u *dirname, int force)
+{
+  char_u      *p;
+
+  if (buf->b_fname != NULL
+      && !bt_nofile(buf)
+      && !path_with_url((char *)buf->b_fname)
+      && (force
+          || buf->b_sfname == NULL
+          || path_is_absolute(buf->b_sfname))) {
+    xfree(buf->b_sfname);
+    buf->b_sfname = NULL;
+    p = path_shorten_fname(buf->b_ffname, dirname);
+    if (p != NULL) {
+      buf->b_sfname = vim_strsave(p);
+      buf->b_fname = buf->b_sfname;
+    }
+    if (p == NULL || buf->b_fname == NULL)
+      buf->b_fname = buf->b_ffname;
+  }
+}
+
+/*
+ * Shorten filenames for all buffers.
+ */
 void shorten_fnames(int force)
 {
   char_u dirname[MAXPATHL];
-  char_u      *p;
 
   os_dirname(dirname, MAXPATHL);
   FOR_ALL_BUFFERS(buf) {
-    if (buf->b_fname != NULL
-        && !bt_nofile(buf)
-        && !path_with_url((char *)buf->b_fname)
-        && (force
-            || buf->b_sfname == NULL
-            || path_is_absolute(buf->b_sfname))) {
-      xfree(buf->b_sfname);
-      buf->b_sfname = NULL;
-      p = path_shorten_fname(buf->b_ffname, dirname);
-      if (p != NULL) {
-        buf->b_sfname = vim_strsave(p);
-        buf->b_fname = buf->b_sfname;
-      }
-      if (p == NULL || buf->b_fname == NULL)
-        buf->b_fname = buf->b_ffname;
-    }
+      shorten_buf_fname(buf, dirname, force);
 
     /* Always make the swap file name a full path, a "nofile" buffer may
      * also have a swap file. */

--- a/src/nvim/quickfix.c
+++ b/src/nvim/quickfix.c
@@ -2240,6 +2240,9 @@ void qf_list(exarg_T *eap)
     }
   }
 
+  /* Shorten all the file names, so that it is easy to read */
+  shorten_fnames(FALSE);
+
   if (qi->qf_lists[qi->qf_curlist].qf_nonevalid)
     all = TRUE;
   qfp = qi->qf_lists[qi->qf_curlist].qf_start;
@@ -2944,6 +2947,10 @@ static void qf_fill_buffer(qf_info_T *qi, buf_T *buf, qfline_T *old_last)
 
   /* Check if there is anything to display */
   if (qi->qf_curlist < qi->qf_listcount) {
+     char_u dirname[MAXPATHL];
+
+     *dirname = NUL;
+
     // Add one line for each error
     if (old_last == NULL) {
       qfp = qi->qf_lists[qi->qf_curlist].qf_start;
@@ -2959,6 +2966,14 @@ static void qf_fill_buffer(qf_info_T *qi, buf_T *buf, qfline_T *old_last)
         if (qfp->qf_type == 1) {  // :helpgrep
           STRLCPY(IObuff, path_tail(errbuf->b_fname), sizeof(IObuff));
         } else {
+          /* shorten the file name if not done already */
+          if (errbuf->b_sfname == NULL
+              || path_is_absolute(errbuf->b_sfname))
+          {
+            if (*dirname == NUL)
+              os_dirname(dirname, MAXPATHL);
+            shorten_buf_fname(errbuf, dirname, FALSE);
+          }
           STRLCPY(IObuff, errbuf->b_fname, sizeof(IObuff));
         }
         len = (int)STRLEN(IObuff);

--- a/src/nvim/testdir/test_quickfix.vim
+++ b/src/nvim/testdir/test_quickfix.vim
@@ -2584,3 +2584,27 @@ func Test_qf_id()
   call Xqfid_tests('c')
   call Xqfid_tests('l')
 endfunc
+
+" Test for shortening/simplifying the file name when opening the
+" quickfix window or when displaying the quickfix list
+func Test_shorten_fname()
+  if !has('unix')
+    return
+  endif
+  %bwipe
+  " Create a quickfix list with a absolute path filename
+  let fname = getcwd() . '/test_quickfix.vim'
+  call setqflist([], ' ', {'lines':[fname . ":20:Line20"], 'efm':'%f:%l:%m'})
+  call assert_equal(fname, bufname('test_quickfix.vim'))
+  " Opening the quickfix window should simplify the file path
+  cwindow
+  call assert_equal('test_quickfix.vim', bufname('test_quickfix.vim'))
+  cclose
+  %bwipe
+  " Create a quickfix list with a absolute path filename
+  call setqflist([], ' ', {'lines':[fname . ":20:Line20"], 'efm':'%f:%l:%m'})
+  call assert_equal(fname, bufname('test_quickfix.vim'))
+  " Displaying the quickfix list should simplify the file path
+  silent! clist
+  call assert_equal('test_quickfix.vim', bufname('test_quickfix.vim'))
+endfunc

